src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java

```diff
@@ -94,4 +95,107 @@ public void testOffsetAtRead(String cigar, int expectedReferencePos, int posInRe
+
+        deepCopy = sam.deepCopy();
+
+        // test for attribute reference (in)equality and content equality
+        final byte samBytes[] = sam.getByteArrayAttribute("BY");
+        final byte copyBytes[] = deepCopy.getByteArrayAttribute("BY");
+        Assert.assertFalse(copyBytes == bytes);
+        Assert.assertFalse(copyBytes == samBytes);
+        Assert.assertTrue(Arrays.equals(copyBytes, samBytes));
+
+        final short samShorts[] = sam.getSignedShortArrayAttribute("SH");
+        final short copyShorts[] = deepCopy.getSignedShortArrayAttribute("SH");
+        Assert.assertFalse(copyShorts == shorts);
+        Assert.assertFalse(copyShorts == samShorts);
+        Assert.assertTrue(Arrays.equals(copyShorts, samShorts));
+
+        final  int samInts[] = sam.getSignedIntArrayAttribute("IN");
+        final  int copyInts[] = deepCopy.getSignedIntArrayAttribute("IN");
+        Assert.assertFalse(copyInts == ints);
+        Assert.assertFalse(copyInts == samInts);
+        Assert.assertTrue(Arrays.equals(copyInts, samInts));
+
+        final float samFloats[] = sam.getFloatArrayAttribute("FL");
+        final float copyFloats[] = deepCopy.getFloatArrayAttribute("FL");
+        Assert.assertFalse(copyFloats == floats);
+        Assert.assertFalse(copyFloats == samFloats);
+        Assert.assertTrue(Arrays.equals(copyFloats, samFloats));
+
+        // now unsigned...
+        sam.setUnsignedArrayAttribute("BY", bytes);
+        sam.setUnsignedArrayAttribute("SH", shorts);
+        sam.setUnsignedArrayAttribute("IN", ints);
+
+        deepCopy = sam.deepCopy();
+
+        final byte samUBytes[] = sam.getUnsignedByteArrayAttribute("BY");
+        final byte copyUBytes[] = deepCopy.getUnsignedByteArrayAttribute("BY");
+        Assert.assertFalse(copyUBytes == bytes);
+        Assert.assertFalse(copyUBytes == samUBytes);
+        Assert.assertTrue(Arrays.equals(copyUBytes, samUBytes));
+
+        final short samUShorts[] = sam.getUnsignedShortArrayAttribute("SH");
+        final short copyUShorts[] = deepCopy.getUnsignedShortArrayAttribute("SH");
+        Assert.assertFalse(copyUShorts == shorts);
+        Assert.assertFalse(copyUShorts == samUShorts);
+        Assert.assertTrue(Arrays.equals(copyUShorts, samUShorts));
+
+        final int samUInts[] = sam.getUnsignedIntArrayAttribute("IN");
+        final int copyUInts[] = deepCopy.getUnsignedIntArrayAttribute("IN");
+        Assert.assertFalse(copyUInts == ints);
+        Assert.assertFalse(copyUInts == samUInts);
+        Assert.assertTrue(Arrays.equals(copyUInts, samUInts));
+    }
+
+    private SAMRecord testDeepCopy(SAMRecord sam) {
+        SAMRecord deepCopy = sam.deepCopy();
+
+        // force the indexing bins to be computed in order to satisfy equality test
+        sam.setIndexingBin(sam.computeIndexingBin());
+        deepCopy.setIndexingBin(deepCopy.computeIndexingBin());
+        Assert.assertTrue(sam.equals(deepCopy));
+
+        return deepCopy;
+    }
```

Comment:
Add a test case that deep-copies a `SAMRecord`, then mutates some non-attributes (ideally both a reference and non-reference field -- eg., a String field and a primitive field), and shows that the original record is unchanged.